src/index/MetaDataTypes.h

```diff
@@ -0,0 +1,189 @@
+#include <vector>
+
+#include "../global/Id.h"
+#include "../util/File.h"
+#include "../util/HashMap.h"
+
+using std::array;
+using std::pair;
+using std::vector;
+
+// Check index_layout.md for explanations (expected comments).
+// Removed comments here so that not two places had to be kept up-to-date.
+//
+// (joka921): After restructuring this file only contains the basic single
+// building blocks of the metaData for one relation.
+
+static const uint64_t IS_FUNCTIONAL_MASK = 0x0100000000000000;
+static const uint64_t HAS_BLOCKS_MASK = 0x0200000000000000;
+static const uint64_t NOF_ELEMENTS_MASK = 0x000000FFFFFFFFFF;
+static const uint64_t MAX_NOF_ELEMENTS = NOF_ELEMENTS_MASK;
+
+class BlockMetaData {
+ public:
+  BlockMetaData() : _firstLhs(0), _startOffset(0) {}
+
+  BlockMetaData(Id lhs, off_t start) : _firstLhs(lhs), _startOffset(start) {}
+
+  Id _firstLhs;
+  off_t _startOffset;
+};
+
+class FullRelationMetaData {
+ public:
+  FullRelationMetaData();
+  
+
+  FullRelationMetaData(Id relId, off_t startFullIndex, size_t nofElements,
+                       double col1Mult, double col2Mult, bool isFunctional,
+                       bool hasBlocks);
+
+  size_t getNofBytesForFulltextIndex() const;
+
+  // Returns true if there is exactly one RHS for each LHS in the relation
+  bool isFunctional() const;
+
+  bool hasBlocks() const;
+
+  // Handle the fact that those properties are encoded in the same
+  // size_t as the number of elements.
+  void setIsFunctional(bool isFunctional);
+
+  void setHasBlocks(bool hasBlocks);
+
+  void setCol1LogMultiplicity(uint8_t mult);
+  void setCol2LogMultiplicity(uint8_t mult);
+  uint8_t getCol1LogMultiplicity() const;
+  uint8_t getCol2LogMultiplicity() const;
+
+  size_t getNofElements() const;
+
+  // Restores meta data from raw memory.
+  // Needed when registering an index on startup.
+  FullRelationMetaData& createFromByteBuffer(unsigned char* buffer);
+
+  // The size this object will require when serialized to file.
+  size_t bytesRequired() const;
+
+  off_t getStartOfLhs() const;
+
+  Id _relId;
+  off_t _startFullIndex;
+
+  friend ad_utility::File& operator<<(ad_utility::File& f,
+                                      const FullRelationMetaData& rmd);
+
+  // operators needed for checking of emptyness
+  // inequality is the common case, so we implement this
+  bool operator!=(const FullRelationMetaData& other) const {
+    return _relId != other._relId ||_typeMultAndNofElements != other._typeMultAndNofElements || _startFullIndex != other._startFullIndex;
+  }
+
+  // __________________________________________________________________
+  bool operator==(const FullRelationMetaData& other) const {
+    return !(*this != other);
+  }
+ private:
+  // first byte: type
+  // second byte: log(col1Multiplicity)
+  // third byte: log(col2Multiplicity)
+  // other 5 bytes: the nof elements.
+  uint64_t _typeMultAndNofElements;
+};
+
+inline ad_utility::File& operator<<(ad_utility::File& f,
```

Comment:
I'm not sure I like that this uses the `<<` operator, why not just a method like `writeFullRelationMetaData()`?

Reply:
The operator<< are also from the original QLever code which I kept wherever possible. An I think, for simple data types which are written to the current position of a stream they are the c++ way to go.
I think there are two cases:
operator<< - just write a bunch of serialized bytes to a stream
member function writeThisToFile(fileRepresentation) - also involves some kind of file management (opening, closing, or at least moving the file pointer to the end.

Reply:
I see, makes sense